### PR TITLE
feat(`launch`): events

### LINKED ISSUE #427: Investigate on Event emitting for `launch`
Investigate and list events to be emitted for the `launch` module

### PR DESCRIPTION
**Pull request by aljo242**
Closes #427

### What does this PR does?

Add `events.proto` for the `x/launch` module and emit cosmos sdk typed events on key updates.  The events should contain all information necessary for a client or DB to get the state of the chain without running any queries.

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Include tests
- [X] Respect code style and lint
- [ ] Update documentation (*.md) (if needed)

### CODE REVIEW

**Comment by lubtd** [comment]
As discussed, let's keep this feature on hold

**Comment by aljo242** [comment]
> As discussed, let's keep this feature on hold

should we close the PR or keep open?

**Comment by lubtd** [comment]
> should we close the PR or keep open?

Let's just keep it here for now

File: x/launch/keeper/msg_server_create_chain.go [hunk 1]
```diff
@@ -35,7 +36,16 @@ func (k msgServer) CreateChain(goCtx context.Context, msg *types.MsgCreateChain)
 		return nil, sdkerrors.Wrap(types.ErrCreateChainFail, err.Error())
 	}
 
+	chain, found := k.GetChain(ctx, id)
```
Comment (Pantani) [inline x/launch/keeper/msg_server_create_chain.go:39]: I think we don't need to check if the chain exists.

File: x/launch/keeper/msg_server_request_add_account.go [hunk 2]
```diff
@@ -47,19 +47,26 @@ func (k msgServer) RequestAddAccount(
 	}
 
 	var requestID uint64
+	var err error
```
Comment (Pantani) [inline x/launch/keeper/msg_server_request_add_account.go:50]:
```suggestion
	var (
		requestID uint64
		err error
	)
```

File: x/launch/keeper/msg_server_request_remove_account.go [hunk 3]
```diff
@@ -42,6 +42,7 @@ func (k msgServer) RequestRemoveAccount(
 	}
 
 	var requestID uint64
+	var err error
```
Comment (Pantani) [inline x/launch/keeper/msg_server_request_remove_account.go:45]:
```suggestion
	var (
		requestID uint64
		err error
	)
```

**Comment by lubtd** [comment]
Hey @aljo242, we can move forward with this task :) 
Normally the conflicts should be quite generic

File: x/launch/keeper/msg_edit_chain.go [hunk 4]
```diff
@@ -72,6 +71,9 @@ func (k msgServer) EditChain(goCtx context.Context, msg *types.MsgEditChain) (*t
 	}
 
 	k.SetChain(ctx, chain)
+	err = ctx.EventManager().EmitTypedEvent(&types.EventChainUpdated{
```
Comment (giunatale) [inline x/launch/keeper/msg_edit_chain.go:74]: lint: ineffectual assignment to err (ineffassign)

File: x/launch/keeper/msg_revert_launch.go [hunk 5]
```diff
@@ -57,6 +57,9 @@ func (k msgServer) RevertLaunch(goCtx context.Context, msg *types.MsgRevertLaunc
 
 	// clear associated client IDs from monitoring
 	k.monitoringcKeeper.ClearVerifiedClientIDs(ctx, msg.LaunchID)
+	err = ctx.EventManager().EmitTypedEvent(&types.EventLaunchReverted{
```
Comment (giunatale) [inline x/launch/keeper/msg_revert_launch.go:60]: ineffectual assignment to err (ineffassign)

**Review by giunatale — CHANGES_REQUESTED** [review]

**Comment by aljo242** [comment]
CodeCov again :|

File: x/launch/keeper/msg_request_add_validator.go [hunk 6]
```diff
@@ -67,6 +67,9 @@ func (k msgServer) RequestAddValidator(
 
 		err = ctx.EventManager().EmitTypedEvent(&types.EventValidatorAdded{
 			GenesisValidator: *content.GetGenesisValidator(),
+			LaunchID:         msg.LaunchID,
+			HasCampaign:      chain.HasCampaign,
+			CampaignID:       chain.CampaignID,
```
Comment (Matthews3301) [inline x/launch/keeper/msg_request_add_validator.go:72]: Would it be possible to also emit:
1. New validator operator address.
2. The address of the account which approved this validator (coordinator)

**Comment by aljo242** [comment]
@Matthews3301 I have added the `CoordinatorAddress` to the Event.  The validator operator address can be accessed with the `GenesisValidator` field of the event, which is already present.

**Comment by Matthews3301** [comment]
> @Matthews3301 I have added the `CoordinatorAddress` to the Event. The validator operator address can be accessed with the `GenesisValidator` field of the event, which is already present.

Thank you - looks good. Maybe for consistency sake, we should have the same data emitted for `EventValidatorRemoved`?

**Comment by aljo242** [comment]
> > @Matthews3301 I have added the `CoordinatorAddress` to the Event. The validator operator address can be accessed with the `GenesisValidator` field of the event, which is already present.
> 
> Thank you - looks good. Maybe for consistency sake, we should have the same data emitted for `EventValidatorRemoved`?

The only information that could possibly change between "adding" and "removing" is the `CoordinatorAddress`, so I will add that.  Do we need that information for any of the other events?

**Comment by Matthews3301** [comment]
@aljo242 None that I'm aware of. This should be good for now.

File: x/launch/keeper/msg_create_chain.go [hunk 7]
```diff
@@ -50,7 +50,11 @@ func (k msgServer) CreateChain(goCtx context.Context, msg *types.MsgCreateChain)
 		}
 	}
 
+	chain, _ := k.GetChain(ctx, id)
+	err = ctx.EventManager().EmitTypedEvent(&types.EventChainCreated{
```
Comment (lubtd) [inline x/launch/keeper/msg_create_chain.go:54]: In `campaign`, a chain for the mainnet is also created by calling `CreateNewChain` when sending `MsgIninitializeMainnet`

Maybe we could have this event in `CreateNewChain` to catch everything? Or we can eventually consider a `MainnetInitialized` event

File: x/launch/keeper/request.go [hunk 8]
```diff
@@ -131,32 +132,62 @@ func CheckAccount(ctx sdk.Context, k Keeper, launchID uint64, address string) (b
...
+	err := CheckRequest(ctx, k, chain.LaunchID, request)
+	if err != nil {
 		return err
 	}
 
 	switch requestContent := request.Content.Content.(type) {
 	case *types.RequestContent_GenesisAccount:
 		ga := requestContent.GenesisAccount
 		k.SetGenesisAccount(ctx, *ga)
+		err = ctx.EventManager().EmitTypedEvent(&types.EventGenesisAccountAdded{
```
Comment (lubtd) [inline x/launch/keeper/request.go:148]: For all events here, we also emit the event in the message for request when sent by the coordinator. So the event would be sent twice

We should keep only these events here

**Comment by aljo242** [comment]
@lubtd I moved the event emission to `CreateNewChain`

File: x/launch/keeper/request.go [hunk 9]
```diff
@@ -131,32 +132,62 @@ func CheckAccount(ctx sdk.Context, k Keeper, launchID uint64, address string) (b
...
+
 	case *types.RequestContent_GenesisValidator:
 		ga := requestContent.GenesisValidator
 		k.SetGenesisValidator(ctx, *ga)
+		err = ctx.EventManager().EmitTypedEvent(&types.EventValidatorAdded{
+			GenesisValidator:   *ga,
+			LaunchID:           chain.LaunchID,
+			HasCampaign:        chain.HasCampaign,
+			CampaignID:         chain.CampaignID,
+			CoordinatorAddress: coord.Address,
```
Comment (lubtd) [inline x/launch/keeper/request.go:178]: I think we can have `CoordinatorAddress` for every request applied event for consistency

**Review by lubtd — COMMENTED** [review]
Tested with the CLI👍

**Review by lubtd — APPROVED** [review]

**Review by giunatale — APPROVED** [review]

**Pull request merged**